src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,170 @@
     public T get(int index) {
-        return null;
+        return findNode(index) == null ? null : findNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldValue = findNode(index).value;
+        findNode(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> removed = findNode(index);
+        unlink(removed);
+        return removed.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> removed = findNode(object);
+        if (removed != null) {
+            unlink(removed);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value,Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+
+    }
+
+    private void unlink(Node<T> unlinked) {
+        Node<T> nodeBefore = unlinked.prev;
+        Node<T> nodeAfter = unlinked.next;
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else if (nodeBefore == null) {
+            nodeAfter.prev = null;
+            head = nodeAfter;
+        } else if (nodeAfter == null) {
+            nodeBefore.next = null;
+            tail = nodeBefore;
+        } else {
+            nodeAfter.prev = nodeBefore;
+            nodeBefore.next = nodeAfter;
+        }
+        size--;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index is not valid!");
+        }
+    }
+
+    private void checkHead(int index, Node<T> newNode) {
+        head = index == 0 ? newNode : head;
+    }
+
+    private void linkNodes(Node<T> beforeNode, Node<T> newNode, Node<T> afterNode) {
+        if (beforeNode != null) {
+            beforeNode.next = newNode;
+        }
+        if (afterNode != null) {
+            afterNode.prev = newNode;
+        }
+    }
+
+    private Node<T> start(int index) {
+        return index > size / 2 ? tail : head;
+    }
+
+    private Node<T> findNode(int index) {
+        checkIndex(index);
+        Node<T> current;
+
+        if (start(index) == head
+                || start(index) != null && start(index).equals(head)) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
```

Comment:
```suggestion
        checkIndex(index);
        Node<T> current;
        if (start(index) == head
                || start(index) != null && start(index).equals(head)) {
            current = head;
            for (int i = 0; i < index; i++) {
                current = current.next;
            }
            return current;
        } else {
            current = tail;
            for (int i = size - 1; i > index; i--) {
                current = current.prev;
            }
            return current;
        }
```
і після цього можна спростити метод get()